packages/eslint-plugin/src/rules/member-ordering.ts

```diff
@@ -707,34 +820,111 @@ export default util.createRule<Options, MessageIds>({
+          const grouped = checkGroupSort(
+            memberSet,
+            memberTypes,
+            supportsModifiers,
+          );
+
+          if (grouped === null) {
+            return false;
+          }
+
+          if (hasAlphaSort) {
+            return !grouped.some(
+              groupMember =>
+                !checkAlphaSort(groupMember, order as AlphabeticalOrder),
+            );
+          }
+        } else if (hasAlphaSort) {
+          return checkAlphaSort(memberSet, order as AlphabeticalOrder);
+        }
+
+        return true;
+      };
 
       if (Array.isArray(orderConfig)) {
         memberTypes = orderConfig;
       } else {
         order = orderConfig.order;
         memberTypes = orderConfig.memberTypes;
+        required = orderConfig.required;
       }
 
-      const hasAlphaSort = !!(order && order !== 'as-written');
+      if (!checkRequiredOrder(members, required)) {
+        return;
+      }
 
-      // Check order
-      if (Array.isArray(memberTypes)) {
-        const grouped = checkGroupSort(members, memberTypes, supportsModifiers);
+      if (required === 'first') {
+        // if the order of required and optional elements is correct,
+        // then check for correct sort and group order within the required and
+        // optional member sets
+        const firstOptionalMemberIndex = members.findIndex(member =>
+          isMemberOptional(member),
+        );
+        const lastRequiredMemberIndex = findLastIndex(
+          members,
+          m => !isMemberOptional(m),
+        );
 
-        if (grouped === null) {
-          return;
+        if (firstOptionalMemberIndex != -1) {
+          const optionalMembers: Member[] = members.slice(
+            firstOptionalMemberIndex,
+          );
+          memberSets.push(optionalMembers);
         }
 
-        if (hasAlphaSort) {
-          grouped.some(
-            groupMember =>
-              !checkAlphaSort(groupMember, order as AlphabeticalOrder),
+        if (lastRequiredMemberIndex != -1) {
+          const requiredMembers: Member[] = members.slice(
+            0,
+            lastRequiredMemberIndex + 1,
           );
+          memberSets.push(requiredMembers);
         }
-      } else if (hasAlphaSort) {
-        checkAlphaSort(members, order as AlphabeticalOrder);
+      } else if (required === 'last') {
+        // if the order of required and optional elements is correct,
+        // then check for correct order within the required and
+        // optional member sets
+        const firstRequiredMemberIndex = members.findIndex(
+          member => !isMemberOptional(member),
+        );
+        const lastOptionalMemberIndex = findLastIndex<Member>(
+          members,
+          isMemberOptional,
+        );
+
+        if (firstRequiredMemberIndex != -1) {
+          const requiredMembers: Member[] = members.slice(
+            firstRequiredMemberIndex,
+          );
+          memberSets.push(requiredMembers);
```

Comment:
[Nitpick] Why bother making the variable if it's only used once, on a next line?

```suggestion
          memberSets.push(members.slice(firstRequiredMemberIndex));
```

Here and with the other three times this is done.

Reply:
I wanted verbosity over shorter code hence, the explicit variable creation.